flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -350,66 +355,64 @@ private boolean isPartialResourceConfigured(JobGraph jobGraph) {
-
-		jobManagerRunnerFutures.put(jobGraph.getJobID(), jobManagerRunnerFuture);
-
-		return jobManagerRunnerFuture
-			.thenApply(FunctionUtils.uncheckedFunction(this::startJobManagerRunner))
-			.thenApply(FunctionUtils.nullFn())
-			.whenCompleteAsync(
-				(ignored, throwable) -> {
-					if (throwable != null) {
-						jobManagerRunnerFutures.remove(jobGraph.getJobID());
-					}
-				},
-				getMainThreadExecutor());
-	}
+	/**
+	 * @param blocking If true, the returned future will complete once the JobManager has been initialized
+	 */
+	private CompletableFuture<Void> runJob(JobGraph jobGraph, boolean blocking) {
+		Preconditions.checkState(!runningJobs.containsKey(jobGraph.getJobID()));
+		long jobManagerInitializationStarted = System.currentTimeMillis();
+		DispatcherJob dispatcherJob = new DispatcherJob();
+		runningJobs.put(jobGraph.getJobID(), dispatcherJob);
+
+		CompletableFuture<JobManagerRunner> initializingJobManager = createJobManagerRunner(jobGraph);
+		dispatcherJob.setInitializingJobManagerRunnerFuture(initializingJobManager);
+		CompletableFuture<JobMasterGateway> initializingJobMasterGateway = CompletableFuture.supplyAsync(() -> new InitializingJobMasterGateway(initializingJobManager, jobGraph, dispatcherJob),
+			getRpcService().getExecutor());
+		dispatcherJob.setInitializingJobMasterGatewayFuture(initializingJobMasterGateway);
 
-	private CompletableFuture<JobManagerRunner> createJobManagerRunner(JobGraph jobGraph) {
-		final RpcService rpcService = getRpcService();
+		initializingJobManager.thenApplyAsync(
+				FunctionUtils.uncheckedFunction((runner) -> {
+					JobManagerRunner r = startJobManagerRunner(runner);
+					dispatcherJob.setInitializingJobMasterGatewayFuture(null);
+					return r;
+				}),
+				getRpcService().getExecutor()); // execute in separate pool to avoid blocking the Dispatcher
 
-		return CompletableFuture.supplyAsync(
-			() -> {
-				try {
-					return jobManagerRunnerFactory.createJobManagerRunner(
-						jobGraph,
-						configuration,
-						rpcService,
-						highAvailabilityServices,
-						heartbeatServices,
-						jobManagerSharedServices,
-						new DefaultJobManagerJobMetricGroupFactory(jobManagerMetricGroup),
-						fatalErrorHandler);
-				} catch (Exception e) {
-					throw new CompletionException(new JobExecutionException(jobGraph.getJobID(), "Could not instantiate JobManager.", e));
+		initializingJobManager.whenCompleteAsync((ignored, throwable) -> {
+			if (throwable != null) {
+				DispatcherJob job = runningJobs.get(jobGraph.getJobID());
+				boolean isCancelled = false;
+				if (job != null) {
+					isCancelled = job.isCancelled();
 				}
-			},
-			rpcService.getExecutor());
+				// error during initialization
+				onJobManagerInitializationFailure(
+					jobGraph,
+					throwable,
+					jobManagerInitializationStarted,
+					isCancelled);
+			}
+		}, getMainThreadExecutor()); // execute in main thread to avoid concurrency issues
```

Comment:
archiving should not run in the main thread since it involves IO operations.

Reply:
Good point. The reason why I execute this in the main thread is because I'm mutating the `runningJobs` map in that method.
I'll try to execute the cleanup in two phases, using different executors.